ethernet_init/ethernet_init.c

```diff
@@ -0,0 +1,382 @@
+
+#if CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET || CONFIG_EXAMPLE_USE_SPI_ETHERNET
+    ESP_GOTO_ON_FALSE(eth_handles_out != NULL && eth_cnt_out != NULL, ESP_ERR_INVALID_ARG,
+                      err, TAG, "invalid arguments: initialized handles array or number of interfaces");
+    eth_handles_g = calloc(SPI_ETHERNETS_NUM + INTERNAL_ETHERNETS_NUM, sizeof(esp_eth_handle_t));
+    ESP_GOTO_ON_FALSE(eth_handles_g != NULL, ESP_ERR_NO_MEM, err, TAG, "no memory");
+
+#if CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET
+    eth_handles_g[eth_cnt_g] = eth_init_internal(&mac_g[eth_cnt_g], &phy_g[eth_cnt_g]);
+    ESP_GOTO_ON_FALSE(eth_handles_g[eth_cnt_g], ESP_FAIL, err, TAG, "internal Ethernet init failed");
+    eth_cnt_g++;
+#endif //CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET
+
+#if CONFIG_EXAMPLE_USE_SPI_ETHERNET
+    ESP_GOTO_ON_ERROR(spi_bus_init(), err, TAG, "SPI bus init failed");
+    // Init specific SPI Ethernet module configuration from Kconfig (CS GPIO, Interrupt GPIO, etc.)
+    spi_eth_module_config_t spi_eth_module_config[CONFIG_EXAMPLE_SPI_ETHERNETS_NUM] = { 0 };
+    INIT_SPI_ETH_MODULE_CONFIG(spi_eth_module_config, 0);
+
+    // The SPI Ethernet module(s) might not have a burned factory MAC address, hence use manually configured address(es).
+    // In this example, Locally Administered MAC address derived from ESP32x base MAC address is used.
+    // Note that Locally Administered OUI range should be used only when testing on a LAN under your control!
+    uint8_t base_mac_addr[ETH_ADDR_LEN];
+    ESP_GOTO_ON_ERROR(esp_efuse_mac_get_default(base_mac_addr), err, TAG, "get EFUSE MAC failed");
+    uint8_t local_mac_1[ETH_ADDR_LEN];
+    esp_derive_local_mac(local_mac_1, base_mac_addr);
+    spi_eth_module_config[0].mac_addr = local_mac_1;
+
+#if CONFIG_EXAMPLE_SPI_ETHERNETS_NUM > 1
+    INIT_SPI_ETH_MODULE_CONFIG(spi_eth_module_config, 1);
+    uint8_t local_mac_2[ETH_ADDR_LEN];
+    base_mac_addr[ETH_ADDR_LEN - 1] += 1;
+    esp_derive_local_mac(local_mac_2, base_mac_addr);
+    spi_eth_module_config[1].mac_addr = local_mac_2;
+#endif
+
+#if CONFIG_EXAMPLE_SPI_ETHERNETS_NUM > 2
+#error Maximum number of supported SPI Ethernet devices is currently limited to 2 by this example.
+#endif
+
+    for (int i = 0; i < CONFIG_EXAMPLE_SPI_ETHERNETS_NUM; i++) {
+        eth_handles_g[eth_cnt_g] = eth_init_spi(&spi_eth_module_config[i], &mac_g[eth_cnt_g], &phy_g[eth_cnt_g]);
+        ESP_GOTO_ON_FALSE(eth_handles_g[eth_cnt_g], ESP_FAIL, err, TAG, "SPI Ethernet init failed");
+        eth_cnt_g++;
+    }
+#endif // CONFIG_EXAMPLE_USE_SPI_ETHERNET
+
+#else
+    ESP_LOGD(TAG, "no Ethernet device selected to init");
+#endif // CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET || CONFIG_EXAMPLE_USE_SPI_ETHERNET
+    *eth_handles_out = eth_handles_g;
+    *eth_cnt_out = eth_cnt_g;
+
+    return ret;
+#if CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET || CONFIG_EXAMPLE_USE_SPI_ETHERNET
+err:
+    free(eth_handles_g);
+    return ret;
+#endif
+}
+
+
+void ethernet_deinit(void)
```

Comment:
`ethernet_init` returns array of Ethernet handles. Therefore `ethernet_deinit` should take this array as input parameter rather than `void`, in my opinion.

Reply:
Or make `esp_eth_handle_t *eth_handles_out[], uint8_t *eth_cnt_out` optional output parameters of `ethernet_init`.

Reply:
```ethernet_deinit_all()``` now takes eth_handles as input argument.